optimum/habana/transformers/models/mpt/modeling_mpt.py

```diff
@@ -60,21 +64,31 @@ def gaudi_mpt_attention_forward(
     if past_key_value is not None:
         if len(past_key_value) != 0:
             if token_idx is not None:
+                # copy kv to kv_cache
                 past_key_value[0].index_copy_(2, token_idx - 1, key_states)
                 past_key_value[1].index_copy_(2, token_idx - 1, value_states)
-                key_states = past_key_value[0]
-                value_states = past_key_value[1]
+
+                if cache_idx is not None:
+                    key_states = past_key_value[0][:, :, :cache_idx, :]
+                    value_states = past_key_value[1][:, :, :cache_idx, :]
+                else:
+                    key_states = past_key_value[0]
+                    value_states = past_key_value[1]
             else:
                 key_states = torch.cat([past_key_value[0], key_states], dim=2)
                 value_states = torch.cat([past_key_value[1], value_states], dim=2)
                 past_key_value = [key_states, value_states]
     else:
-        past_key_value = [
-            torch.empty(key_states.shape, dtype=key_states.dtype, device=key_states.device),
-            torch.empty(key_states.shape, dtype=key_states.dtype, device=key_states.device),
-        ]
-        past_key_value[0][:] = key_states[:]
-        past_key_value[1][:] = value_states[:]
+        if cache_idx is not None:
+            cache_shape = (batch_size, self.n_heads, kv_cache_len, self.head_dim)
```

Comment:
what is the need for bringing kv_cache_len in? instead of using q_len?

Reply:
I updated the implementation to set kv_cache_len as an attribute of attention class. 
The code logic here is initializing kv_cache. When enable bucket internal, the input will be sliced from 'max_input_tokens+max_output_token' to 'cache_idx', whicle the length of kv_cahce should be 'max_input_tokens+max_output_token', i.e. kv_cache_len.